### ISSUE TITLE
Replace BBCode mess with shortcode lexer

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#4488]
This is a tracking issue for the work ahead. The new shortcode parser will be based on https://github.com/thunderer/Shortcode and will support all the currently supported BBCode tag and perform the auto-linker in a way that won't break everything all the time.

Related issues:
- [] #1493
- [ ] #2199
- [ ] #2330
- [ ] #3218 
- [ ] #3641
- [ ] #4014
- [ ] #4222
- [ ] #4451
- [ ] #4453
- [ ] https://forum.friendi.ca/display/39bbe52a605a8c3b1f8ed7e629667549

**Comment by annando** [comment]
I have one wish: Additional to the current way to integrate pictures I would like to have one where you can define the "description" field. This is important for improved accessibility.

**Comment by MrPetovan** [comment]
This actually is the easiest wish, this library supports shortcodes with parameters like we do with attachments. It would simply be `[img description="Long Description" alt="Alt"]http://example.com/image.png[/img]`.

The two blocking points so far with the standard parser:
- It doesn't recognize `[*]` as a self-closing shortcode for lists
- It doesn't recognize single-parameters shortcodes like this: `[url=http://example.com]link[/url]`

The good news is that we can write our own parser, we don't have to wait for the library maintainer to update it. The bad news is that it's 01:30 already and I work tomorrow. But I'm having fun!

**Comment by annando** [comment]
Coders are creatures of the night ... :-) Sleep well.

**Comment by MrPetovan** [comment]
Waiting on https://github.com/thunderer/Shortcode/issues/65

**Comment by MrPetovan** [comment]
There has been some movement over at https://github.com/thunderer/Shortcode/pull/66 so this is on my plate for the next version.

**Comment by JoshyPHP** [comment]
Hello, I found this issue via a Google alert related to BBCodes. I develop and maintain a library that handles BBCodes as well as other markup and if you haven't committed to an implementation yet, you may want to take a look at it. Here's [the s9e\TextFormatter project page](https://github.com/s9e/TextFormatter) and [the RTD manual](https://s9etextformatter.readthedocs.io/).

The library was designed to handle various forms of markup in a unified way. All of the BBCodes are customizable and the grammar supports multiple attributes, quoted or unquoted. The library has a plugin for Unicode emoji with configurable aliases (e.g. EmojiOne's alpha codes) and a plugin for ASCII-style emoticons/smilies such as `;)`. All text is parsed into an XML representation which [can be stored in your database](https://s9etextformatter.readthedocs.io/Getting_started/How_it_works/) for performance, transformed to HTML or converted back to text. The library uses XSL as templating language, which can be compiled to native PHP.

I am not familiar with Friendica's inner workings but I'm willing to assist with anything related to the library. The simplest way to use the library may be to [create your own bundle](https://s9etextformatter.readthedocs.io/Bundles/Your_own_bundle/) which is basically a serialized version of a fully-configured parser and renderer. I have published an [example project that creates/uses its own bundle](https://github.com/s9e/MyProject), you can take a look at the `BundleConfigurator.php` file for an example and the `www/index.php` file for how to use it.

I'm subscribed to this issue so let me know if you have any questions or if you want to move this discussion to a more appropriate place.

**Comment by annando** [comment]
@JoshyPHP We are using BBCode variations like `[img=http://domain.tld/picture]Alternate Description[/img]` or complete custom codes like `[attachment type='some type' url='https://domain.tld/link/to/article' title='Some title' image='https://domain.tld/link/to/picture.png']Some more description of that link[/attachment]`. Is this possible?

We also do have codes like `[nobb]Some stuff that isn't expected to be rendered at all[/nobb]`. Additionally it should be able to add some code that links that are not encapsulated by `[url]` are transformed into links that use `[url]`. We also are having different "dialects" depending upon the target. Means: When we use HTML to display stuff, we create it slightly different in comparison on when we  transmit it via AP, and so on.

If your parser is able to do such things, then I guess you would make many people here totally happy.

**Comment by MrPetovan** [comment]
Thanks for your offer, I haven't looked into this for a while because of the hold-up related to the third-party library I was looking into and other stuff came up.

My main requirement for a BBCode parser and the reason I was looking into a lexer is that I want it to identify "free text" that can be subject to smilies transformation and URL autolinking. This issue lists the related issues that I think would be solved by having such an identification.

**Comment by JoshyPHP** [comment]
@annando Yes, all of that seems possible. I whipped up a quick example to try it out. BBCode definitions use a short [custom syntax](https://s9etextformatter.readthedocs.io/Plugins/BBCodes/Custom_BBCode_syntax/) inspired by phpBB's (the library is used in phpBB 3.2+) but you can also define them using a more verbose, object-oriented API.
```php
$configurator = new s9e\TextFormatter\Configurator;

$configurator->Autolink;
$configurator->BBCodes->addCustom(
	'[img src={URL} alt={TEXT;useContent}]',
	'<img src="{@src}" alt="{@alt}"/>'
);
$configurator->BBCodes->addCustom(
	'[attachment type={TEXT} url={URL} title={TEXT} image={URL} desc={TEXT;useContent}]{TEXT}[/attachment]',
	'<div title="{@title}"><img src="{@image}" alt="{@desc}"/></div>'
);
$configurator->BBCodes->addCustom(
	'[nobb]{TEXT}[/nobb]',
	'<pre><code>{TEXT}</code></pre>'
);

extract($configurator->finalize());

$text = "[img=http://domain.tld/picture]Alternate Description[/img]\n"
      . "[attachment type='some type' url='https://domain.tld/link/to/article' title='Some title' image='https://domain.tld/link/to/picture.png']Some more description of that link[/attachment]\n"
      . "[nobb][img=http://domain.tld/picture]Alternate Description[/img][/nobb]\n"
      . 'https://example.org';
$xml  = $parser->parse($text);
$html = $renderer->render($xml);

die("$html\n");
```
```html
<img src="http://domain.tld/picture" alt="Alternate Description">
<div title="Some title"><img src="https://domain.tld/link/to/picture.png" alt="Some more description of that link"></div>
<pre><code>[img=http://domain.tld/picture]Alternate Description[/img]</code></pre>
<a href="https://example.org">https://example.org</a>
```
In the example above, you may notice that the `[nobb]` BBCode is simply defined as a code block and we don't explicitly set it to ignore markup. This is because the library analyses templates and [automatically generates rules](https://s9etextformatter.readthedocs.io/Rules/Automatic_rules_generation/) accordingly. One of those rules says that text inside of a code block should not be formatted, e.g. no BBCodes, no links, etc... That's for convenience and those rules can be selectively disabled.

There are multiple ways to change the HTML output depending on the target. If it's a simple change then you can use a conditional in the template itself, in the form `<xsl:if then="$TARGET = 'foo'>`. You can also toggle plugins and individual tags, but depending on how different the configuration is you may want to simply create a different parser and/or renderer for each target.


@MrPetovan I went over the list of issues linked to this one and most of them are fairly common and should be resolvable without too much effort. Compared to something like `thunderer/Shortcode` I'd say the library is less "hands-on" and you don't get as involved with the parsing. You don't have to execute as much of your own PHP during parsing and you can't execute PHP at all during rendering. One way to use dynamically-generated data is to modify the XML representation between parsing and rendering. For example, that's [how phpBB uses localized dates in user quotes](https://github.com/phpbb/phpbb/blob/480016f600ce13e5fecf4c0ab82349de436a2897/phpBB/phpbb/textformatter/s9e/quote_helper.php#L47-L80). Although if you plan to parse and render user content every time it's displayed then you can also do that during parsing.

Another difference that may matter is that the current version of the library (2.x) requires PHP 7.1. I saw in Friendica's requirements that it supports PHP 7.0.

**Comment by MrPetovan** [comment]
Thank you for the elaboration, however I'm not comfortable with having additional intermediate artifacts, whether it is the XML or the generated bundle. The BBCode text already is the intermediate artifact, and ideally it should be converted to the display HTML.

**Comment by JoshyPHP** [comment]
There's an easy fix for that: don't use the XML representation. The primary reason it exists is for efficiency; Instead of reparsing the same text on every page, it makes it possible to parse it only once and still render it dynamically without losing the original text representation. It also allows storing metadata (OpenGraph data and that kind of things) or the result of costly operations as part of the document. If efficiency isn't a priority, you can parse and render the text in a single operation.

There's another benefit to having access to the parsed text in XML form: you can manipulate it with DOM. Take for example the code block from #7808. The XML would look like this:
```xml
<r><CODE><s>[code]</s>[guid] =&gt; <e>[/code]</e></CODE></r>
```
Once loaded in a DOM, there's no ambiguity about where the BBCode markup is (in `s` and `e` elements) and what the content of the block is without having to encode/decode it. Similarly, in forum softwares it makes is easy to remove excessive blockquotes at arbitrary nesting depth by selecting which elements to remove via XPath in a single, safe operation. Or it makes it possible to find user mentions via XPath rather than doing it manually via regular expressions that may not stay in sync with the ones used in the parser.

If you feel like having no rules and/or configuring everything by hand you can clear them all with `$configurator->rulesGenerator->clear()`. In my experience, it's rarely (never?) desirable that's why the library uses sensible defaults instead of initializing everything in a blank state.

**Comment by MrPetovan** [comment]
Thanks for the extra information.
